File: sharedmodule/llmswitch-core/src/router/virtual-router/load-balancer.ts
import type { LoadBalancingPolicy } from './types.js';

type StickyRouteEntry = {
  providerKey: string;
  updatedAtMs: number;
};

interface RouteState {
  pointer: number;
  stickyMap: Map<string, StickyRouteEntry>;
  weighted: {
    currentWeights: Map<string, number>;
  };
}

export interface LoadBalancingOptions {
  routeName: string;
  candidates: string[];
  stickyKey?: string;
  weights?: Record<string, number>;
  availabilityCheck: (providerKey: string) => boolean;
}

export interface GroupedLoadBalancingOptions {
  routeName: string;
  groups: Map<string, string[]>;
  stickyKey?: string;
  weights?: Record<string, number>;
  availabilityCheck: (providerKey: string) => boolean;
}

const DEFAULT_STICKY_TTL_MS = 30 * 60_000;
const DEFAULT_STICKY_MAX_ENTRIES = 4096;

function resolveStickyTtlMs(): number {
  const raw = process.env.ROUTECODEX_VR_STICKY_TTL_MS ?? process.env.RCC_VR_STICKY_TTL_MS;
  const parsed = Number.parseInt(String(raw ?? '').trim(), 10);
  if (Number.isFinite(parsed) && parsed >= 60_000) {
    return parsed;
  }
  return DEFAULT_STICKY_TTL_MS;
}

function resolveStickyMaxEntries(): number {
  const raw = process.env.ROUTECODEX_VR_STICKY_MAX_ENTRIES ?? process.env.RCC_VR_STICKY_MAX_ENTRIES;
  const parsed = Number.parseInt(String(raw ?? '').trim(), 10);
  if (Number.isFinite(parsed) && parsed >= 16) {
    return parsed;
  }
  return DEFAULT_STICKY_MAX_ENTRIES;
}

export class RouteLoadBalancer {
  private policy: LoadBalancingPolicy;
  private readonly states: Map<string, RouteState> = new Map();
  private readonly stickyTtlMs: number;
  private readonly stickyMaxEntries: number;

  constructor(policy?: LoadBalancingPolicy) {
    this.policy = policy ?? { strategy: 'round-robin' };
    this.stickyTtlMs = resolveStickyTtlMs();
    this.stickyMaxEntries = resolveStickyMaxEntries();
  }

  updatePolicy(policy?: LoadBalancingPolicy): void {
    if (policy) {
      this.policy = policy;
    }
  }

  getPolicy(): LoadBalancingPolicy {
    return this.policy;
  }

  select(options: LoadBalancingOptions, strategyOverride?: LoadBalancingPolicy['strategy']): string | null {
    const available = options.candidates.filter((candidate) => options.availabilityCheck(candidate));
    if (available.length === 0) {
      return null;
    }

    const strategy = strategyOverride ?? this.policy.strategy;
    switch (strategy) {
      case 'sticky':
        return this.selectSticky(options.routeName, available, options.stickyKey, options.weights ?? this.policy.weights);
      case 'weighted':
        return this.selectWeighted(options.routeName, available, options.weights ?? this.policy.weights);
      default:
        if (options.weights) {
          const distinct = new Set(available.map((candidate) => normalizeWeight(options.weights?.[candidate])));
          if (distinct.size > 1) {
            return this.selectWeighted(options.routeName, available, options.weights);
          }
        }
        return this.selectRoundRobin(options.routeName, available);
    }
  }

  selectGrouped(
    options: GroupedLoadBalancingOptions,
    strategyOverride?: LoadBalancingPolicy['strategy']
  ): string | null {
    const groupIds = Array.from(options.groups.keys()).filter((groupId) => {
      const members = options.groups.get(groupId) ?? [];
      return members.some((candidate) => options.availabilityCheck(candidate));
    });
    if (groupIds.length === 0) {
      return null;
    }

    const normalizedWeights = normalizeGroupWeights(groupIds, options.weights);
    const strategy = strategyOverride ?? this.policy.strategy;
    const groupRoute = `${options.routeName}:group`;

    let selectedGroup: string | null = null;
    switch (strategy) {
      case 'sticky':
        selectedGroup = this.selectSticky(groupRoute, groupIds, options.stickyKey, normalizedWeights);
        break;
      case 'weighted':
        selectedGroup = this.selectWeighted(groupRoute, groupIds, normalizedWeights);
        break;
      default:
        selectedGroup = this.selectRoundRobin(groupRoute, groupIds);
        break;
    }

    if (!selectedGroup) {
      return null;
    }
    const groupCandidates = (options.groups.get(selectedGroup) ?? []).filter((candidate) =>
      options.availabilityCheck(candidate)
    );
    if (groupCandidates.length === 0) {
      return null;
    }
    return this.selectRoundRobin(`${groupRoute}:${selectedGroup}`, groupCandidates);
  }

  private selectRoundRobin(routeName: string, candidates: string[]): string {
    const state = this.getState(routeName);
    const choice = candidates[state.pointer % candidates.length];
    state.pointer = (state.pointer + 1) % candidates.length;
    return choice;
  }

  private selectWeighted(routeName: string, candidates: string[], weights?: Record<string, number>): string {
    // Deterministic smooth weighted round-robin (no randomness) so routing behavior is testable and stable.
    // Each candidate with a positive weight is guaranteed to be selected eventually.
    const state = this.getState(routeName);
    const current = state.weighted.currentWeights;

    const candidateSet = new Set(candidates);
    for (const existing of Array.from(current.keys())) {
      if (!candidateSet.has(existing)) {
        current.delete(existing);
      }
    }
    for (const key of candidates) {
      if (!current.has(key)) {
        current.set(key, 0);
      }
    }

    const candidateWeights = candidates.map((candidate) => normalizeWeight(weights?.[candidate]));
    const totalWeight = candidateWeights.reduce((sum, w) => sum + w, 0);
    if (!Number.isFinite(totalWeight) || totalWeight <= 0) {
      return this.selectRoundRobin(routeName, candidates);
    }

    let bestIndex = 0;
    let bestScore = Number.NEGATIVE_INFINITY;
    for (let i = 0; i < candidates.length; i += 1) {
      const key = candidates[i];
      const w = candidateWeights[i];
      const next = (current.get(key) ?? 0) + w;
      current.set(key, next);
      if (next > bestScore) {
        bestScore = next;
        bestIndex = i;
      }
    }

    const selectedKey = candidates[bestIndex];
    current.set(selectedKey, (current.get(selectedKey) ?? 0) - totalWeight);
    return selectedKey;
  }

  private selectSticky(routeName: string, candidates: string[], stickyKey?: string, weights?: Record<string, number>): string {
    if (!stickyKey) {
      return this.selectRoundRobin(routeName, candidates);
    }
    const state = this.getState(routeName);
    const nowMs = Date.now();
    this.pruneStickyEntries(state, nowMs);
    const pinned = state.stickyMap.get(stickyKey);
    if (pinned && candidates.includes(pinned.providerKey)) {
      pinned.updatedAtMs = nowMs;
      return pinned.providerKey;
    }
    const choice =
      weights && Object.keys(weights).length > 0
        ? this.selectWeighted(`${routeName}:sticky`, candidates, weights)
        : this.selectRoundRobin(routeName, candidates);
    state.stickyMap.set(stickyKey, {
      providerKey: choice,
      updatedAtMs: nowMs
    });
    this.pruneStickyEntries(state, nowMs);
    return choice;
  }

  private pruneStickyEntries(state: RouteState, nowMs: number): void {
    for (const [key, entry] of state.stickyMap.entries()) {
      if (!entry || typeof entry.providerKey !== 'string' || nowMs - entry.updatedAtMs >= this.stickyTtlMs) {
        state.stickyMap.delete(key);
      }
    }
    if (state.stickyMap.size <= this.stickyMaxEntries) {
      return;
    }
    const sorted = Array.from(state.stickyMap.entries()).sort((a, b) => a[1].updatedAtMs - b[1].updatedAtMs);
    while (state.stickyMap.size > this.stickyMaxEntries && sorted.length > 0) {
      const [oldestKey] = sorted.shift()!;
      state.stickyMap.delete(oldestKey);
    }
  }

  private getState(routeName: string): RouteState {
    if (!this.states.has(routeName)) {
      this.states.set(routeName, { pointer: 0, stickyMap: new Map(), weighted: { currentWeights: new Map() } });
    }
    return this.states.get(routeName)!;
  }
}

function normalizeWeight(value: number | undefined): number {
  if (typeof value === 'number' && Number.isFinite(value) && value > 0) {
    return value;
  }
  return 1;
}

function normalizeGroupWeights(groupIds: string[], weights?: Record<string, number>): Record<string, number> | undefined {
  if (!weights || groupIds.length === 0) {
    return undefined;
  }
  const normalized: Record<string, number> = {};
  let total = 0;
  for (const id of groupIds) {
    const raw = weights[id];
    const value = typeof raw === 'number' && Number.isFinite(raw) && raw > 0 ? raw : 0;
    if (value > 0) {
      normalized[id] = value;
      total += value;
    }
  }
  if (!Number.isFinite(total) || total <= 0) {
    return undefined;
  }
  for (const id of Object.keys(normalized)) {
    normalized[id] = normalized[id] / total;
  }
  return normalized;
}

File: sharedmodule/llmswitch-core/rust-core/crates/router-hotpath-napi/src/virtual_router_engine/load_balancer.rs
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

use super::context_weighted::ContextWeightedConfig;
use super::health_weighted::HealthWeightedConfig;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct LoadBalancingPolicy {
    pub strategy: Option<String>,
    pub weights: Option<HashMap<String, i64>>,
    pub health_weighted: Option<HealthWeightedConfig>,
    pub context_weighted: Option<ContextWeightedConfig>,
}

impl Default for LoadBalancingPolicy {
    fn default() -> Self {
        Self {
            strategy: Some("round-robin".to_string()),
            weights: None,
            health_weighted: None,
            context_weighted: None,
        }
    }
}

#[derive(Debug, Default, Clone)]
struct RouteState {
    pointer: usize,
    weighted_current: HashMap<String, i64>,
}

#[derive(Debug, Default, Clone)]
pub(crate) struct RouteLoadBalancer {
    policy: LoadBalancingPolicy,
    states: HashMap<String, RouteState>,
}

impl RouteLoadBalancer {
    pub(crate) fn new(policy: Option<LoadBalancingPolicy>) -> Self {
        Self {
            policy: policy.unwrap_or(LoadBalancingPolicy {
                strategy: Some("round-robin".to_string()),
                weights: None,
                health_weighted: None,
                context_weighted: None,
            }),
            states: HashMap::new(),
        }
    }

    pub(crate) fn update_policy(&mut self, policy: Option<LoadBalancingPolicy>) {
        if let Some(policy) = policy {
            self.policy = policy;
        }
    }

    pub(crate) fn policy(&self) -> &LoadBalancingPolicy {
        &self.policy
    }

    pub(crate) fn select(
        &mut self,
        route_name: &str,
        candidates: &[String],
        _sticky_key: Option<&str>,
        weights: Option<&HashMap<String, i64>>,
        availability_check: impl Fn(&str) -> bool,
        strategy_override: Option<&str>,
    ) -> Option<String> {
        let available: Vec<String> = candidates
            .iter()
            .filter(|candidate| availability_check(candidate.as_str()))
            .cloned()
            .collect();
        if available.is_empty() {
            return None;
        }

        let strategy = strategy_override
            .or_else(|| self.policy.strategy.as_deref())
            .unwrap_or("round-robin");
        let default_weights = self.policy.weights.clone();
        let resolved_weights = weights.or(default_weights.as_ref());
        match strategy {
            "weighted" => Some(self.select_weighted(route_name, &available, resolved_weights)),
            _ => {
                if let Some(custom_weights) = weights {
                    let mut distinct: HashSet<i64> = HashSet::new();
                    for candidate in &available {
                        let weight = custom_weights.get(candidate).cloned().unwrap_or(1).max(1);
                        distinct.insert(weight);
                    }
                    if distinct.len() > 1 {
                        return Some(self.select_weighted(
                            route_name,
                            &available,
                            Some(custom_weights),
                        ));
                    }
                }
                Some(self.select_round_robin(route_name, &available))
            }
        }
    }

    pub(crate) fn select_round_robin_with_skips(
        &mut self,
        route_name: &str,
        candidates: &[String],
        availability_check: impl Fn(&str) -> bool,
    ) -> Option<String> {
        if candidates.is_empty() {
            return None;
        }
        let state = self.get_state_mut(route_name);
        let total = candidates.len();
        for offset in 0..total {
            let idx = (state.pointer + offset) % total;
            let candidate = candidates.get(idx)?;
            if !availability_check(candidate) {
                continue;
            }
            state.pointer = (idx + 1) % total;
            return Some(candidate.clone());
        }
        None
    }

    pub(crate) fn peek_round_robin_with_skips(
        &mut self,
        route_name: &str,
        candidates: &[String],
        availability_check: impl Fn(&str) -> bool,
    ) -> Option<String> {
        if candidates.is_empty() {
            return None;
        }
        let state = self.get_state_mut(route_name);
        let total = candidates.len();
        for offset in 0..total {
            let idx = (state.pointer + offset) % total;
            let candidate = candidates.get(idx)?;
            if availability_check(candidate) {
                return Some(candidate.clone());
            }
        }
        None
    }

    pub(crate) fn select_grouped(
        &mut self,
        route_name: &str,
        ordered_group_ids: &[String],
        groups: &HashMap<String, Vec<String>>,
        _sticky_key: Option<&str>,
        weights: Option<&HashMap<String, i64>>,
        availability_check: impl Fn(&str) -> bool,
        strategy_override: Option<&str>,
    ) -> Option<String> {
        let available_groups: Vec<String> = ordered_group_ids
            .iter()
            .filter(|group_id| {
                groups
                    .get(*group_id)
                    .map(|members| {
                        members
                            .iter()
                            .any(|candidate| availability_check(candidate.as_str()))
                    })
                    .unwrap_or(false)
            })
            .cloned()
            .collect();
        if available_groups.is_empty() {
            return None;
        }

        let strategy = strategy_override
            .or_else(|| self.policy.strategy.as_deref())
            .unwrap_or("round-robin");
        let group_route = format!("{}:group", route_name);
        let selected_group = match strategy {
            "weighted" => self.select_weighted(&group_route, &available_groups, weights),
            _ => self.select_round_robin(&group_route, &available_groups),
        };

        let group_candidates: Vec<String> = groups
            .get(&selected_group)?
            .iter()
            .filter(|candidate| availability_check(candidate.as_str()))
            .cloned()
            .collect();
        if group_candidates.is_empty() {
            return None;
        }

        Some(self.select_round_robin(
            &format!("{}:{}", group_route, selected_group),
            &group_candidates,
        ))
    }

    pub(crate) fn select_round_robin_for_route(
        &mut self,
        route_name: &str,
        candidates: &[String],
    ) -> Option<String> {
        if candidates.is_empty() {
            return None;
        }
        Some(self.select_round_robin(route_name, candidates))
    }

    fn select_round_robin(&mut self, route_name: &str, candidates: &[String]) -> String {
        let state = self.get_state_mut(route_name);
        let idx = state.pointer % candidates.len();
        let choice = candidates[idx].clone();
        state.pointer = (state.pointer + 1) % candidates.len();
        choice
    }

    fn select_weighted(
        &mut self,
        route_name: &str,
        candidates: &[String],
        weights: Option<&HashMap<String, i64>>,
    ) -> String {
        let state = self.get_state_mut(route_name);
        let current = &mut state.weighted_current;
        let candidate_set: HashSet<String> = candidates.iter().cloned().collect();
        current.retain(|key, _| candidate_set.contains(key));
        for key in candidates {
            current.entry(key.clone()).or_insert(0);
        }

        let candidate_weights: Vec<i64> = candidates
            .iter()
            .map(|candidate| {
                weights
                    .and_then(|w| w.get(candidate))
                    .cloned()
                    .unwrap_or(1)
                    .max(1)
            })
            .collect();
        let total_weight: i64 = candidate_weights.iter().sum();

        let mut best_index = 0usize;
        let mut best_score: i64 = i64::MIN;
        for (idx, candidate) in candidates.iter().enumerate() {
            let weight = candidate_weights[idx];
            let next = current.get(candidate).cloned().unwrap_or(0) + weight;
            current.insert(candidate.clone(), next);
            if next > best_score {
                best_score = next;
                best_index = idx;
            }
        }
        let selected_key = candidates[best_index].clone();
        let current_value = current.get(&selected_key).cloned().unwrap_or(0) - total_weight;
        current.insert(selected_key.clone(), current_value);
        selected_key
    }

    fn get_state_mut(&mut self, route_name: &str) -> &mut RouteState {
        self.states
            .entry(route_name.to_string())
            .or_insert_with(RouteState::default)
    }
}

#[cfg(test)]
mod tests {
    use super::*;


    #[test]
    fn sticky_strategy_is_treated_as_round_robin_for_non_continuation_requests() {
        let mut lb = RouteLoadBalancer::new(Some(LoadBalancingPolicy {
            strategy: Some("sticky".to_string()),
            weights: None,
            health_weighted: None,
            context_weighted: None,
        }));
        let candidates = vec!["provider-a".to_string(), "provider-b".to_string()];

        let first = lb
            .select("thinking", &candidates, Some("session:same"), None, |_| true, None)
            .expect("first selection");
        let second = lb
            .select("thinking", &candidates, Some("session:same"), None, |_| true, None)
            .expect("second selection");

        assert_eq!(first, "provider-a");
        assert_eq!(second, "provider-b", "non-continuation requests must not be pinned by sticky load balancing");
    }

    #[test]
    fn round_robin_covers_all_aliases_in_single_model_pool() {
        let mut lb = RouteLoadBalancer::new(None);
        let candidates = (1..=5)
            .map(|idx| format!("windsurf.ws-pro-{idx}.gpt-5.4-none"))
            .collect::<Vec<_>>();
        let mut hits = Vec::new();
        for _ in 0..5 {
            hits.push(
                lb.select_round_robin_for_route("thinking", &candidates)
                    .expect("candidate should be selected"),
            );
        }
        assert_eq!(hits, candidates);
    }
}
